rfd/0xxx-tls-routing-behind-lb.md

```diff
@@ -0,0 +1,286 @@
+  │                       │                            │
+  │                       ├──┐                         │
+  │                       │  │generate:                │
+  │                       │  │local CA                 │
+  │                       │  │local client credentials │
+  │                       │◄─┘local KUBECONFIG         │
+  │                       │                            │
+  │ server localhost:8888 │                            │
+  │ local credentials     │                            │
+  ├──────────────────────►│  server kube-xxx.xxx.xxx   │
+  │                       │  "real" credentials        │
+  │                       ├───────────────────────────►│
+  │                       │                            │
+  │                       │◄───────────────────────────┤
+  │◄──────────────────────┤                            │
+  │                       │                            │
+  │ another request       │                            │
+  ├──────────────────────►│                            |
+  │                       ├───────────────────────────►│
+  ...                   
+```
+
+Once the local server receives a request and verifies the request, the local server will proxy the request using the TLS
+Routing dialer to connect the Teleport Proxy. The local proxy should be in charge of managing the Teleport cert used for
+routing Kubernetes requests.
+
+Note that `kube-teleport-proxy-alpn.my-teleport-cluster.com` is not required to be resolvable (by DNS) in this case, but
+it will be used internally for routing.
+
+### User Experience
+
+Once the connection upgrade support to all protocols are implemented, users can be recommended to upgrade their Teleport
+clusters to single port mode if the cluster is currently serving separate ports.
+
+All scenarios discussed in this section assumes Teleport sits behind a load balancer so connection upgrade for TLS
+Routing is required.
+
+#### 1 - Database Access UX
+
+Common flows already use `tsh` local proxy (since v8) so they will automatically perform the connection upgrade:
+
+- `tsh db connect` (single port mode)
+- `tsh proxy db`
+
+So no UX change to these.
+
+In single port mode, using native clients to connect to Teleport Proxy directly with `tsh db env/config` will NOT work
+through load balancers. Users can use one of the above methods instead.
+
+#### 2 - Application Access UX
+
+Application Access through Teleport Webapp is NOT affected by the issue of interest of this RFD.
+
+On the `tsh` side, TCP, AWS, Azure, and GCP apps always use a local proxy so they will automatically perform the
+connection upgrade when needed. No UX change to these.
+
+However, HTTP apps MUST use `tsh proxy app`.
+
+#### 3 - Kubernetes Access UX
+
+Using native Kubernetes clients after `tsh kube login` will NOT work as the native clients cannot upgrade the
+connection.
+
+`tsh kubectl` will automatically starts a local proxy and performs the connection upgrade when needed.
```

Comment:
Do we plan to support standalone kubectl command ? 
I think at least  in case of running `kubectl` without kubeproxy we should consider printing description message that the `kubectl` requires running local `tsh proxy kube`

Reply:
Added explanation that `tsh kube login` should advertise `tsh xxxx`. I am considering in this case, we probably SHOULDN'T even update the default kube config like `~/.kube/config`.

Reply:
We can support vanilla `kubectl` if the user exposes a new copy of the kubeconfig when using `tsh proxy kube .... -p ....`-

We have customers using `go`, `js`, and `python` clients so we must keep compatibility with them. `tsh proxy kube ...` must create a deep copy of the user's default kubeconfig and edit the required fields. Once it terminates, it cleans the kubeconfig so even if the user tries to use it, kubectl will fail because the file does not exist. The only thing the user must do is configure the `KUBECONFIG` env var in the new shell.

`tsh kubectl` can start the proxy, login, and create the new kubeconfig automatically for simplicity